WyattHWilliams/react-challenge-project-jan-2022
# #6 Add-Feature: Route Guarding

## Changes
1. created a RequireLogin element
2. applied requireLogin to appRouter
3. refactored login functionality pipeline to asynchronous code to await getting a login response from the server and set that response to the authReducer before the requireLogin element checks for authentication

## Purpose
to protect routes from being accessible to a non-logged in user.
Addresses [issue 6](https://github.com/Shift3/react-challenge-project-jan-2022/issues/6)

## Approach
1. requireLogin checks the authReducer to the user's email information. if none is present, it kicks the user back to the login page.
2. requireLogin runs whenever any non-/login route is requested.
3. When the user is logging in, we want to wait for the server to evaluate the login info and send us a response before the requireLogin makes its check, otherwise requireLogin will kick the user back to login just before the authReducer records the users login info. In this case, requireLogin will kick the user back to login even if the server returns a successful response. to avoid this, I changed the login functionality to run asynchronously and wait for a response to be processed from the server and logged to the authReducer before the requireLogin does it's evaluation.

## Pre-Testing TODOs
1. be logged out.

## Testing Steps
1. access /login, make sure that works!
2. try to access the protected routes without logging in, check that user is redirected to /login
3. login
4. test accessing all the routes, check that each is accessed properly.
5. log the user out
6. try to access the protected routes without logging in, check that user is redirected to /login

Closes #006
